cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,197 @@
+   const int w_id = tid / warp_size;
+   // begin/end indices for the column data
+   size_type begin = 0;
+   size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   size_type w_begin = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(begin);
+   size_type w_end = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(end);
+
+   // lane id within the current warp
+   constexpr size_type w_leader_lane{0};
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   size_type w_valid_count{0};
+  
+   // current warp.
+   size_type w_cur = w_begin + w_id; 
+   auto w_active_mask = __ballot_sync(0xFFFF'FFFF, w_cur <= w_end);
+   while(w_cur <= w_end){
+      // absolute element index
+      size_type index = (w_cur * warp_size) + w_lane_id;
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && filter(index) ? lhs.is_valid(index) : rhs.is_valid(index);
+      }
+
+      // do the copy if-else, but only if this element is valid in the column to be copied 
+      if(in_range && valid){ 
+         out.element<T>(index) = filter(index) ? lhs.element<T>(index) : rhs.element<T>(index);
+      }
+      
+      // update validity
+      if(has_validity){         
+         // the final validity mask for this warp
+         int w_mask = __ballot_sync(w_active_mask, valid);
+         // only one guy in the warp needs to update the mask and count
+         if(w_lane_id == 0){
+            out.set_mask_word(w_cur, w_mask);            
+            w_valid_count += __popc(w_mask);
+         }
+      }      
+
+      // next grid
+      w_cur += blockDim.x * gridDim.x;
+      if(has_validity){
+         w_active_mask = __ballot_sync(w_active_mask, w_cur < w_end);
+      }
+   }
+
+   if(has_validity){
+      // sum all null counts across all warps
+      size_type b_valid_count = single_lane_block_sum_reduce<block_size, w_leader_lane>(w_valid_count);
+      // b_null_count will only be valid on thread 0
+      if(threadIdx.x == 0){
+         // using an atomic here because there are multiple blocks doing this work
+         atomicAdd(valid_count, b_valid_count);
+      }      
+   }
+}
+
+
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief Functor called by the `type_dispatcher` in order to perform a copy if/else
+*        using a filter function to select from lhs/rhs columns.
+*/
+/* ----------------------------------------------------------------------------*/
+struct copy_if_else_functor {
+   template <typename T, typename Filter>
+   void operator()(  column_view const& lhs,
+                     column_view const& rhs,
+                     Filter filter,
+                     mutable_column_view& out,
```

Comment:
This function should return the output column instead of accepting a mutable view. See below for the reason why.